controller/cache/cluster.go

```diff
@@ -0,0 +1,228 @@
+package cache
+
+import (
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/argoproj/argo-cd/common"
+	appv1 "github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/util"
+	"github.com/argoproj/argo-cd/util/kube"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+const (
+	clusterSyncTimeout = 1 * time.Hour
+)
+
+type clusterInfo struct {
+	nodes        map[string]*node
+	lock         *sync.Mutex
+	onAppUpdated func(appName string)
+	kubectl      kube.Kubectl
+	cluster      *appv1.Cluster
+	syncLock     *sync.Mutex
+	syncTime     *time.Time
+}
+
+func createObjInfo(un *unstructured.Unstructured) *node {
+	ownerRefs := un.GetOwnerReferences()
+	// Special case for endpoint. Remove after https://github.com/kubernetes/kubernetes/issues/28483 is fixed
+	if un.GetKind() == kube.EndpointsKind && len(un.GetOwnerReferences()) == 0 {
```

Comment:
Also check for group in addition to Kind (in case Endpoint CRD is installed).

Reply:
Done: https://github.com/argoproj/argo-cd/pull/840/commits/a842eb3bb31f378c37866ea5deba3efc150d4341